### DISCUSSION TITLE
Cannot build Concourse after adding new dependency

### DISCUSSION THREAD

**Discussion by spire-allyjweir** [concourse/concourse#6091] [General]
Hi there :wave:,

I'm working on a PR to introduce Honeycomb as a tracing exporter option. I've added the necessary code and updated the dependencies but now when I try to build Concourse itself via Docker Compose, it fails on an unrelated step,

```
➜  concourse git:(add-honeycomb-tracing) docker-compose up --build                 
Building web
Step 1/19 : FROM concourse/dev
 ---> 1e4309eccd71
Step 2/19 : WORKDIR /src
 ---> Using cache
 ---> c16dfbc7b876
Step 3/19 : COPY go.mod .
 ---> Using cache
 ---> 4ed0c8039d59
Step 4/19 : COPY go.sum .
 ---> Using cache
 ---> d9b53a2da62a
Step 5/19 : RUN grep '^replace' go.mod || go mod download
 ---> Running in 3243278d187f
replace github.com/docker/distribution v2.7.1+incompatible => github.com/docker/distribution v2.7.1-0.20190205005809-0d3efadf0154+incompatible
replace github.com/jessevdk/go-flags => github.com/vito/go-flags v1.4.1-0.20200428200343-c7161c3bd74d
Removing intermediate container 3243278d187f
 ---> 0638d7488ca0
Step 6/19 : COPY . .
 ---> c1a223afe43b
Step 7/19 : RUN go build -gcflags=all="-N -l" -o /usr/local/concourse/bin/concourse       ./cmd/concourse
 ---> Running in add3cd83c859
go: downloading github.com/honeycombio/opentelemetry-exporter-go v0.2.1
go: downloading github.com/honeycombio/libhoney-go v1.12.0
go: downloading github.com/google/uuid v1.1.1
go: downloading gopkg.in/alexcesaro/statsd.v2 v2.0.0
go: downloading github.com/vmihailenco/msgpack v4.0.4+incompatible
go: downloading github.com/facebookgo/muster v0.0.0-20150708232844-fd3d7953fd52
go: downloading github.com/facebookgo/limitgroup v0.0.0-20150612190941-6abd8d71ec01
go: downloading github.com/facebookgo/clock v0.0.0-20150410010913-600d898af40a
Removing intermediate container add3cd83c859
 ---> 4d2083b11212
Step 8/19 : VOLUME /src
 ---> Running in 30d8bb03b34f
Removing intermediate container 30d8bb03b34f
 ---> 103f523ed0b2
Step 9/19 : RUN  set -x &&      gcc -O2 -static -o /usr/local/concourse/bin/init ./cmd/init/init.c
 ---> Running in 4801de148b2e
+ gcc -O2 -static -o /usr/local/concourse/bin/init ./cmd/init/init.c
Removing intermediate container 4801de148b2e
 ---> 651769d20400
Step 10/19 : RUN mkdir -p /concourse-keys
 ---> Running in 0498e9998b11
Removing intermediate container 0498e9998b11
 ---> 5cf7106b62ce
Step 11/19 : RUN concourse generate-key -t rsa -b 1024 -f /concourse-keys/session_signing_key
 ---> Running in 20dc1f8ca747
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x331a1fa]

goroutine 1 [running]:
main.QuickstartCommand.LessenRequirements(0x0, 0x0, 0x0)
        /src/cmd/concourse/quickstart.go:25 +0x3a
main.ConcourseCommand.LessenRequirements(0x3a07f88, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /src/cmd/concourse/command.go:27 +0x95
main.main()
        /src/cmd/concourse/main.go:23 +0x145
ERROR: Service 'web' failed to build: The command '/bin/sh -c concourse generate-key -t rsa -b 1024 -f /concourse-keys/session_signing_key' returned a non-zero code: 2
```

The commit with all my changes is here: https://github.com/spire-allyjweir/concourse/commit/1cee886b0409a1c59e463709ed6967b9063afbb0

I don't use Go regularly so I'm worried this issue might be a common dependency issue or 'gotcha' that golang developers might be able to spot easily.

Any help appreciated!

**Comment by taylorsilva** [comment]
This was a fun little challenge!
tldr: In Go, bools always default to `false`. Therefore you don't need to specify `default:"false"` which apparently the `go-flags` package doesn't accept as a valid. A bit of a nit on the package's part 🤷 
Change `Debug` to:
```patch
-Debug       bool   `long:"honeycomb-debug" description:"honeycomb.io debug mode enabled" default:"false"`
+Debug       bool   `long:"honeycomb-debug" description:"honeycomb.io debug mode enabled"`
```

---

The flag parser does throw an error:
```
(dlv) p parser.internalError
error(*github.com/jessevdk/go-flags.Error) *{
	Type: ErrInvalidTag (14),
	Message: "boolean flag `honeycomb-debug' may not have default values, they...+52 more",}
```
but it doesn't return the error until we call `Parse()` or `ParseArgs()`. We don't call either of those funcs right away because we have to wire up our dynamic flags which causes us to enter **the danger zone**. If we could check for errors before trying to wire up our dynamic flags then you would not have run into this problem. Sadly the go-flags package doesn't give us a way to check for errors before calling `Parse()` or a safe way to wire up dynamic flags.

**Reply by jamieklassen** [reply]
nice catch 😎 -- maybe next year we'll switch to a more idiomatic (and more active? I just noticed our PR https://github.com/jessevdk/go-flags/pull/334 got merged months ago lol) flag library as part of https://github.com/concourse/concourse/projects/90#card-39155081

**Comment by spire-allyjweir** [comment]
Thank you very much @taylorsilva! I could tell the usage of `go-flags` wasn't std lib but couldn't tell which library it was part of. Appreciate you taking the time to help get me unblocked.
